packages/lib/src/__tests__/input.test.ts

```diff
@@ -0,0 +1,97 @@
+import { NoLogLogger } from '@smartthings/core-sdk/dist/logger'
+
+import { FileInputProcessor, formatFromFilename, IOFormat, parseJSONOrYAML } from '../input'
+import { LogManager } from '../logger'
+
+
+const resourcesDir = './src/__tests__/resources'
+
+interface SimpleType {
+	num: number
+	str: string
+}
+
+const validYAML = 'num: 5\nstr: my string'
+const validData = { num: 5, str: 'my string' }
+
+describe('formatFromFilename', () => {
+	it('handles yaml extensions', function() {
+		expect(formatFromFilename('fn.yaml')).toBe(IOFormat.YAML)
+		expect(formatFromFilename('fn.yml')).toBe(IOFormat.YAML)
+		expect(formatFromFilename('fn.YAML')).toBe(IOFormat.YAML)
+	})
+
+	it('handles json extensions', function() {
+		expect(formatFromFilename('fn.json')).toBe(IOFormat.JSON)
+		expect(formatFromFilename('fn.JSON')).toBe(IOFormat.JSON)
+	})
+
+	it('defaults to YAML', function() {
+		jest.spyOn(LogManager.prototype, 'getLogger').mockImplementation(() => {
+			return new NoLogLogger()
+		})
+		expect(formatFromFilename('fn')).toBe(IOFormat.YAML)
+	})
+})
+
+describe('parseJSONOrYAML', () => {
+	it('returns data from valid yaml', function() {
+		const result = parseJSONOrYAML(validYAML, 'source')
+
+		expect(result).toEqual(validData)
+	})
+
+	it('throws error with no data', function() {
+		expect(() => parseJSONOrYAML('', 'empty')).toThrow('did not get any data from empty')
+	})
+
+	it('throws error with simple string data', function() {
+		expect(() => parseJSONOrYAML('just a string', 'empty')).toThrow('got simple string from empty')
+	})
+})
+
+// TODO: readDataFromStdin
+/*
+describe('readDataFromStdin', () => {
+	it('returns data from valid yaml', async function() {
+		const buf = Buffer.from(validYAML, 'utf-8')
+		jest.spyOn(process.stdin, 'on')
+			.mockRejectedValueOnce(buf)
+		expect(readDataFromStdin()).resolves.toBe(validYAML)
+	})
+
+	it('passes on error', function() {
+		expect(() => parseJSONOrYAML('', 'empty')).toThrow('did not get any data from empty')
+	})
+})
+*/
+
+describe('FileInputProcessor', () => {
+	it('gets format from filename', function() {
+		// TODO: this really should spy on formatFromFilename
+		// const spy = jest.spyOn(???, 'formatFromFilename')
+		// (I tried unsuccessfully using `input` with `import * as input from '../input' for `???`.)
```

Comment:
Glad you lean that way because that's what I've done and I like how it turned out.

Reply:
Just read that article...had come across it before and stopped reading when I got to "babel" since we don't use babel. TypeScript compiler does the same thing though as it turns out and now I understand something I didn't before. 😄